### PR TITLE
_7zz: 22.01 -> 23.01

### PR DESCRIPTION
**Pull request by eclairevoyant**
## Description of changes

Bump _7zz version, also change `rec` to `finalAttrs` idiom, remove upstreamed patches, and mildly improve update script 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by eclairevoyant** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>_7zz</li>
    <li>peazip</li>
  </ul>
</details>

**Review by Henry-Hiles — APPROVED** [review]
LGTM

**Comment by peterhoeg** [comment]
Doesn't seem to work on darwin according to the tests. Does it currently?

**Comment by eclairevoyant** [comment]
I wouldn't know, I don't own a mac. Listed platforms doesn't indicate darwin support either.

**Comment by peterhoeg** [comment]
`platforms.unix' includes darwin. Additionally, there is a `fix-build-on-darwin.patch' patch so at least at *some* point it worked. If 22.01 is fine on darwin, we need that fixed before this is merged.

**Comment by eclairevoyant** [comment]
All the patches have been applied upstream. But I cannot say why it fails, as I have no way of testing this on a mac, nor would I know if it worked in the past.

File: pkgs/tools/archivers/7zz/default.nix
Comment (06kellyjac) [inline pkgs/tools/archivers/7zz/default.nix:113]: this with is simple and common

**Review by 06kellyjac — CHANGES_REQUESTED** [review]
thanks

has cross compile for mingw also been confirmed to still work?

cc: @AtnNn

File: pkgs/tools/archivers/7zz/default.nix
Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix:113]: "common" is irrelevant, and it's unnecessary.

Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix:113]: I did not double-check cross compilation, will do so now.

**Comment by eclairevoyant** [comment]
`nix build '.#legacyPackages.x86_64-linux.pkgsCross.mingwW64._7zz'` seems to compile fine, not sure how to test functionality atm

File: pkgs/tools/archivers/7zz/default.nix
Comment (06kellyjac) [reply pkgs/tools/archivers/7zz/default.nix:113]: common is **completely** relevant to maintaining a codebase with fellow contributors

And what do you mean unnecessary. Removing this with requires 4 references to `lib.` when it could just be left as it is.

If having to write out 4 copies of `lib.` is fine why is `with` allowed for `maintainers` when it's "unnecessary" you can just do

`maintainers = [ lib.maintainers.anna328p lib.maintainers.eclairevoyant lib.maintainers.jk lib.maintainers.peterhoeg ];`


The downside to `with` are that there's no clarity as to where the value came from when there are multiple, but that's the only `with` up to that point so completely fine

Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix:113]: common is completely irrelevant to improving code quality. and no committer has complained about this change in the past.

Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix:113]: Honestly, every reviewer seems to have their own ideas of what is acceptable, so I'm not interested on bikeshedding a cosmetic change. If you have feedback on the actual build process etc I'll discuss that.

Comment (06kellyjac) [reply pkgs/tools/archivers/7zz/default.nix:113]: please revert the cosmetic change then

Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix:113]: I'd prefer if there is some unified documentation about best practices for nixpkgs. If you can point to that, I will gladly make the change.

Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix]: Alright, as https://nixos.org/manual/nixpkgs/stable/#chap-meta uses `with lib;` I will stick with that. Let the record show that I'm not interested in un-reverting, then.

Comment (06kellyjac) [reply pkgs/tools/archivers/7zz/default.nix]: As I mentioned it's the common pattern across nixpkgs

```
λ rg "meta = with lib; \{" . --stats -q | rg -e "\d+ matched lines"
21126 matched lines

λ rg "meta = \{" . --stats -q | rg -e "\d+ matched lines"
12683 matched lines

# There are a couple files you may want to filter out of the with lib search but very few
λ rg "meta = with lib; \{" . --count | sed 's/:/ /g' | sort -nr -k2 | head -n20
./pkgs/servers/http/nginx/modules.nix 49
./pkgs/applications/editors/eclipse/plugins.nix 29
./pkgs/applications/misc/octoprint/plugins.nix 22
./pkgs/development/libraries/hunspell/dictionaries.nix 13
./pkgs/applications/editors/jetbrains/default.nix 12
./pkgs/servers/search/elasticsearch/plugins.nix 9
./pkgs/misc/tmux-plugins/default.nix 7
./doc/languages-frameworks/python.section.md 7
./pkgs/development/tools/continuous-integration/buildbot/plugins.nix 6
./pkgs/development/python-modules/textx/default.nix 6
./pkgs/development/libraries/unixODBCDrivers/default.nix 6
./pkgs/applications/video/vdr/plugins.nix 6
./pkgs/applications/networking/znc/modules.nix 6
./pkgs/applications/office/activitywatch/default.nix 5
./pkgs/top-level/lua-packages.nix 4
./pkgs/development/tools/replay-io/default.nix 4
./pkgs/data/fonts/noto-fonts/default.nix 4
./pkgs/build-support/setup-hooks/patch-rc-path-hooks/default.nix 4
./pkgs/applications/editors/kakoune/plugins/overrides.nix 4
./pkgs/top-level/perl-packages.nix 3

# For raw meta there are many generated files
λ rg "meta = \{" . --count | sed 's/:/ /g' | sort -nr -k2 | head -n20
./pkgs/development/lisp-modules/imported.nix 4598
./pkgs/top-level/perl-packages.nix 1827
./pkgs/development/lisp-modules-new-obsolete/imported.nix 856
./pkgs/applications/editors/emacs/elisp-packages/elpa-devel-generated.nix 406
./pkgs/development/node-packages/node-packages.nix 396
./pkgs/applications/editors/emacs/elisp-packages/elpa-generated.nix 396
./pkgs/applications/editors/vscode/extensions/default.nix 244
./pkgs/applications/editors/emacs/elisp-packages/nongnu-generated.nix 211
./pkgs/development/lua-modules/generated-packages.nix 111
./pkgs/applications/emulators/retroarch/cores.nix 90
./pkgs/tools/misc/graylog/plugins.nix 20
./pkgs/development/compilers/elm/packages/node-packages.nix 15
./pkgs/applications/science/electronics/bitscope/packages.nix 8
./pkgs/misc/tmux-plugins/default.nix 7
./pkgs/development/python-modules/fenics/default.nix 6
./pkgs/development/ocaml-modules/janestreet/old.nix 6
./pkgs/applications/editors/vim/plugins/overrides.nix 6
./pkgs/applications/networking/browsers/firefox/packages.nix 5
./pkgs/top-level/release-haskell.nix 4
./pkgs/development/libraries/aspell/dictionaries.nix 4

# This is the effect of the filter
λ rg "meta = \{" . -g '!*generated*' -g '!*lisp-modules*' -g '!*perl-packages.nix' -g '!*node-packages.nix' -g '!*vscode' --count  | sed 's/:/ /g' | sort -nr -k2 | head -n20
./pkgs/applications/emulators/retroarch/cores.nix 90
./pkgs/tools/misc/graylog/plugins.nix 20
./pkgs/applications/science/electronics/bitscope/packages.nix 8
./pkgs/misc/tmux-plugins/default.nix 7
./pkgs/development/python-modules/fenics/default.nix 6
./pkgs/development/ocaml-modules/janestreet/old.nix 6
./pkgs/applications/editors/vim/plugins/overrides.nix 6
./pkgs/applications/networking/browsers/firefox/packages.nix 5
./pkgs/top-level/release-haskell.nix 4
./pkgs/development/libraries/aspell/dictionaries.nix 4
./pkgs/build-support/setup-hooks/patch-rc-path-hooks/test/default.nix 4
./pkgs/os-specific/linux/g15daemon/default.nix 3
./pkgs/development/interpreters/bats/libraries.nix 3
./pkgs/applications/virtualization/xen/packages.nix 3
./pkgs/applications/networking/znc/modules.nix 3
./pkgs/applications/graphics/gimp/plugins/default.nix 3
./pkgs/top-level/haxe-packages.nix 2
./pkgs/tools/nix/nixos-render-docs/src/tests/test_lists.py 2
./pkgs/stdenv/linux/make-bootstrap-tools.nix 2
./pkgs/servers/search/manticoresearch/default.nix 2

# and these are the filtered results
λ rg "meta = \{" . -g '!*generated*' -g '!*lisp-modules*' -g '!*perl-packages.nix' -g '!*node-packages.nix' -g '!*vscode' --stats -q | rg -e "\d+ matched lines"
3597 matched lines

# even with that filter and a bunch of extra bits that likely trim out plenty of non-generated results `with lib;` still has it's huge amount of matches
λ rg "meta = with lib; \{" . -g '!*generated*' -g '!*lisp-modules*' -g '!*perl-packages.nix' -g '!*node-packages.nix' -g '!*vscode' -g '!*nginx' -g '!*eclipse' -g '!*octoprint' -g '!*hunspell' -g '!*jetbrains' -g '!*elasticsearch' --stats -q | rg -e "\d+ matched lines"
20974 matched lines

# This is still almost 6 times as many matches
```

There's also this example in the nixpkgs documentation for how to write a meta

https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/meta.chapter.md#meta-attributes-chap-meta

---

A maintainer commenting on a change should not be responded with scare quotes and dismissing the comment immediately as irrelevant.

You also shouldn't bring a cosmetic change and then dismiss all discussion of that change as bikeshedding.

As the person raising the change you should provide documentation and standards for nixpkgs before demanding them from others.

**Review by 06kellyjac — APPROVED** [review]
LGTM, thanks

I don't think we need to wait for aarch64-darwin to build on ofborg or test the mingw cross compile here but we probably should for the backport PR

File: pkgs/tools/archivers/7zz/default.nix
Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix]: My concern is that there is
* **little to no documentation** other than looking at other PRs
* on other PRs committers ask for the exact opposite change
* as mentioned before, "common" does not matter, because there is obviously a lot of bad practices across `nixpkgs` and yet we do not dismiss improvements (and this sentence, paraphrased, has been told to contributors on other PRs, in a far more condescending manner, as I have observed.)

Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix]: Also, I just cited the exact same link and reverted the change already, as mentioned above.

File: pkgs/tools/archivers/7zz/default.nix
Comment (drupol) [inline pkgs/tools/archivers/7zz/default.nix:58]: Could you please add some comments on the reason why this patch is needed ? It is a patch that we can upstream ? When should we remove it?

More info: https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-patch-comment.md

File: pkgs/tools/archivers/7zz/default.nix
Comment (drupol) [inline pkgs/tools/archivers/7zz/default.nix:122]: Thanks for adding yourself as maintainer!

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, let me know if this is clear enough.

Tip for next time, split the `finalAttrs` refactoring in one single commit, then the version update.

File: pkgs/tools/archivers/7zz/default.nix
Comment (eclairevoyant) [reply pkgs/tools/archivers/7zz/default.nix:59]: Not sure how you ran `nixpkgs-hammering`, as it's broken with `finalAttrs` and known as such for over a year (jtojnar/nixpkgs-hammering#131). Anyway comment added

Commits have also been split as requested

Comment (drupol) [reply pkgs/tools/archivers/7zz/default.nix:59]: Actually, I don't use it. I just know the all the defined rules in the project and I like the explanations document... so I use it when I need to justify a feedbackk

**Review by drupol — APPROVED** [review]
LGTM ! So sad we cannot upstream the new patch.

Also, thanks for splitting the commits.

**Comment by drupol** [comment]
I just notice that there's a build issue on `aarch64-darwin`.

I guess it's an extra semicolon.

```
> ../../../../C/AesOpt.c:812:21: fatal error: empty expression statement has no effect; remove unnecessary ';' to silence this warning [-Wextra-semi-stmt]
       >     CTR_START (m, 0);
       >                     ^
       > 1 error generated.
```

**Comment by eclairevoyant** [comment]
Thanks, no need to ping about every error. I have been following the issues reported by ofborg and fixing the points where it errors, force-pushing, repeat (as you can see from the history over the past few days), but I just don't have a way of compiling it on my end (as mentioned above, I don't have a Mac system, so I'm relying on ofborg to report errors related to darwin).

If you have a better alternative to this process, please feel free to share.

**Comment by eclairevoyant** [comment]
The other option I can think of would be to drop the patch and disable the individual warning

**Comment by drupol** [comment]
No problem, no problem, I let you handle the thing, ping me when you want a new review... If you want you can set your PR in draft until it's fixed so we don't make reviews in the meantime.

**Comment by drupol** [comment]
yes

**Comment by eclairevoyant** [comment]
@drupol I believe everything should be good to go now, aarch64-darwin build passes.

**Review by drupol — APPROVED** [review]
Amazing, thanks!

Tip: next time, you can summon  @NixOS/darwin-maintainers if you want some help with Darwin builds and arch.

**Pull request merged**
